### PR TITLE
 gnrc_ipv6: handle hop-by-hop option before forwarding

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Hop-by-hop options need to be handled separately, as they [are required to be handled hop-by-hop](https://tools.ietf.org/html/rfc7045#section-2.2), as the name implies. Now after weeks of rework, we are now finally able to do this cleanly within `gnrc_ipv6`.

### Testing procedure
Since there is no proper hop-by-hop option support yet, I repeated my tests from #10234, but also tested with duplicate Hop-by-hop header, as this is an error case now, as per [RFC 8200](https://tools.ietf.org/html/rfc8200#section-4.1) and activated DEBUG output to verify that it is indeed handled as an error.

### Issues/PRs references
Depends on ~~#8594~~ (merged) and ~~#10234 (and its dependencies)~~ (merged).

PR dependencies

### CODE REVIEW

**Comment by miri64** [comment]
Rebased to current #10234.

**Comment by miri64** [comment]
Already found some errors due to my tests :-)

**Comment by miri64** [comment]
Rebased to current #10234.

**Comment by miri64** [comment]
Rebased to current master and dependencies (and added #8594 as a dependency and updated accordingly).

**Comment by miri64** [comment]
Rebased to current #10244.

**Comment by miri64** [comment]
I reworked this PR to require less `#ifdef`s. Since this PR is not under review yet, I squashed these changes immediately. If your interested in what I did, have look at the [force-push log](https://github.com/RIOT-OS/RIOT/compare/84a6e5a7df3eb96b0904ed9ea24732531a7bbef9..7140adca00ed2b7c3141076ab0fbad48ca177626).

File: sys/include/net/gnrc/ipv6/ext.h [hunk 1]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ *          hop-by-hop header.
+ *
+ * @note    Should be called by @ref net_gnrc_ipv6 after @ref
+ *          gnrc_ipv6_ext_process_hopopt(), i.e. the (first) hop-by-hop option
+ *          should already be marked in @p pkt. If a hop-by-hop option is found
+ *          in gnrc_pktsnip_t::data of @p pkt, the function will return an
+ *          error.
+ *
+ * @param[in] pkt   An IPv6 packet in receive order.
+ * @param[in] nh    **In:** The @ref net_protnum of gnrc_pktsnip_t::data of
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]: I had to re-read and cross-reference this description several times before I (think I) got it. Maybe it could be reworded to something more verbose like

    * @param[in] nh    **In:** The first next header in the extension header
    *                  chain, represented by the @ref net_protnum of the
    *                  gnrc_pktsnip_t::data contained in @p pkt. <br />
    *                  **Out:** The next protocol to be processed (since the
    *                  extension headers are marked now), represented by the
    *                  @ref net_protnum of the gnrc_pktsnip_t::data contained
    *                  in @p pkt.

also, shouldn't it say `@param[in, out] nh` then?

File: sys/include/net/gnrc/ipv6/ext.h [hunk 2]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ *                  @p pkt. <br />
+ *                  **Out:** The new @ref net_protnum of gnrc_pktsnip_t::data of
+ *                  @p pkt (after marking of the extension headers).
+ *
+ * @return  @p pkt with all extension headers marked until the first
+ *           non-extension header.
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]: one whitespace too many (indentation error)

File: sys/include/net/gnrc/ipv6/ext.h (hunk 1 above)
Comment (Lotterleben) [reply sys/include/net/gnrc/ipv6/ext.h]: (verboseness nitpick may also apply to line 68)

File: sys/include/net/gnrc/ipv6/ext.h [hunk 3]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ * @param[in] pkt       An IPv6 packet in receive order. It assumed that
+ *                      gnrc_pktsnip_t::data points to a hop-by-hop option when
+ *                      @p nh points to value equal to
+ *                      @ref PROTNUM_IPV6_EXT_HOPOPT.
+ * @param[in,out] nh    **In:** The @ref net_protnum of gnrc_pktsnip_t::data of
+ *                      @p pkt. <br />
+ *                      **Out:** If @p nh was pointing to a value equal
+ *                      to @ref PROTNUM_IPV6_EXT_HOPOPT on in and the return
+ *                      value is not NULL it points to the next header field
+ *                      of the processed hop-by-hop option header, if the return
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]: very long (and hard to parse) sentence. I'd suggest ending the sentence after `header` and then starting a new one for the "otherwise" option

File: sys/include/net/gnrc/ipv6/ext.h [hunk 4]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ *                      value is NULL the value undefined.
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]: missing "is"

```suggestion
 *                      value is NULL the value is undefined.
```

Comment (Lotterleben) [reply sys/include/net/gnrc/ipv6/ext.h]: Also, "if the value is [...] the value is undefined" is a tad confusing, I'd specify *which* value remains undefined by saying something like `value is NULL the @p nh value is undefined`

File: sys/include/net/gnrc/ipv6/ext.h [hunk 5]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ *                      If @p nh was pointing to a value unequal to @ref
+ *                      PROTNUM_IPV6_EXT_HOPOPT on in the value remains
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]: same value-value confusion as in line 74, consider also adding a `@p nh`

File: sys/include/net/gnrc/ipv6/ext.h [hunk 6]
```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
...
+ *                      unchanged.
+ *
+ * @return  @p pkt with the hop-by-hop option marked on success.
+ * @return  NULL, if packet.was consumed by the option handling.
```
Comment (Lotterleben) [inline sys/include/net/gnrc/ipv6/ext.h]:
```suggestion
 * @return  NULL, if packet was consumed by the option handling.
```

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 7]
```diff
@@ -25,53 +30,174 @@
...
+ * If @p nh is @ref PROTNUM_IPV6_EXT_HOPOPT, the function sends a parameter
+ * problem message (if @ref net_gnrc_icmpv6_error is available), releases the
+ * packet with error `EINVAL`, and returns `false`.
+ *
+ * @note    This function should only be called *after* the first hop-by-hop
+ *          option header in the packet was parsed (if it exists) or if the
+ *          first extension header wasn't a hop-by-hop option header.
+ *
+ * @param[in] pkt   A packet in receive order to be checked with the current
+ *                  extension header (of type @p nh) being the payload.
```
Comment (Lotterleben) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: "as its" instead of "being the"?

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 8]
```diff
@@ -25,53 +30,174 @@
...
+                DEBUG("ipv6: handle extension header (nh = %u)\n", *nh);
+                ext_hdr = pkt->data;
+                if ((pkt = _demux(pkt, *nh)) == NULL) {
+                    DEBUG("ipv6: packet was consumed by extension header "
+                          "handling\n");
+                    return NULL;
+                }
+                *nh = ext_hdr->nh;
+                if (_duplicate_hopopt(pkt, *nh)) {
+                    return NULL;
```
Comment (Lotterleben) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: The documentation of `@param[in] nh ` of `gnrc_ipv6_ext_process_all()` in `ext.h` does not contain a remark about this case (in contrast to the `gnrc_ipv6_ext_process_hopopt()` doc, which says ` if the return value is NULL the value undefined`). Is this on purpose?

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 9]
```diff
@@ -25,53 +30,174 @@
...
     }
-    else {
-        ipv6 = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_IPV6);
-        hdr = ipv6->data;
+    /* remove L2 headers around IPV6 */
+    netif_snip = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_NETIF);
+    if (netif_snip != NULL) {
+        pkt = gnrc_pktbuf_remove_snip(pkt, netif_snip);
+    }
+    pkt = gnrc_pktbuf_reverse_snips(pkt);
```
Comment (Lotterleben) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: mayyyybe add a comment on why they need to be reversed?

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 10]
```diff
@@ -25,53 +30,174 @@
...
+    if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_IPV6,
+                                   GNRC_NETREG_DEMUX_CTX_ALL,
+                                   pkt)) {
+        DEBUG("ipv6_ext_rh: could not dispatch packet to the IPv6 "
+              "thread\n");
+        gnrc_pktbuf_release(pkt);
+    }
+}
 
+static int _handle_rh(gnrc_pktsnip_t *pkt)
```
Comment (Lotterleben) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: tiny comment on what this is for?

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 11]
```diff
@@ -85,53 +211,25 @@ static int _handle_rh(gnrc_pktsnip_t *current, gnrc_pktsnip_t *pkt)
 #endif  /* MODULE_GNRC_IPV6_EXT_RH */
 
 /**
- * @brief marks IPv6 extension header if needed.
- *   updates pkt and returns next header.
- * @param[in] current  The current header
- * @param[in,out] pkt  The whole packet
- * @return The next header
- * @return NULL on error
+ * @brief       Marks IPv6 extension header according to the given length
```
Comment (Lotterleben) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: length given by what? pkt.data.len?

File: sys/include/net/gnrc/ipv6/ext.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]:
```
*                  chain, represented by the @ref net_protnum of the
*                  gnrc_pktsnip_t::data contained in @p pkt. <br />
```

But that's wrong. It's the protocol number of the last parsed header, which isn't in `gnrc_pktsnip_t::data` in `pkt` at this point but in `pkt->next`. But I think such descriptions make it localized confusing as well, since you need to know how packet order works in GNRC.

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: > also, shouldn't it say `@param[in, out] nh` then?

Yepp, sorry will fix.

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: > But that's wrong. It's the protocol number of the last parsed header, which isn't in `gnrc_pktsnip_t::data` in `pkt` at this point but in `pkt->next`. But I think such descriptions make it localized confusing as well, since you need to know how packet order works in GNRC.

Also it isn't true, that `*nh` has to come from the packet either. From a testing perspective it would be totally alright to be some extension header related protocol number. It could just point to `PROTNUM_IPV6_EXT_FRAG` and `pkt->data` will be handled as a fragmentation header.

**Comment by miri64** [comment]
Some remarks you gave are not for this PR, but I will fix them ASAP.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 10 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: Sincet this function is merged with `gnrc_ipv6_ext_rh_process()` #10238, I'd rather not.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 9 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: I will add a comment in #10238.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 8 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: Done

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 11 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: I will fix in the corresponding PR (have to look and see which one that is myself ;-)).

**Comment by miri64** [comment]
Addressed comments that apply to this PR.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 9 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: See https://github.com/RIOT-OS/RIOT/pull/10229/commits/971f1d0fa07010187f79c2616ccf1253c4bced3c (and https://github.com/RIOT-OS/RIOT/pull/10229#issuecomment-441188292).

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 11 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: See https://github.com/RIOT-OS/RIOT/pull/10234/commits/92e8b9f39330b5c71ea70e6bc18f025f9b72a48e (and https://github.com/RIOT-OS/RIOT/pull/10234#issuecomment-441191236).

File: sys/include/net/gnrc/ipv6/ext.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: I hope the way I reformulated the comment is now better understandable.

**Comment by miri64** [comment]
Rebased to current master and dependencies.

**Comment by miri64** [comment]
Rebased to current master. No longer dependent on other PRs.

**Comment by miri64** [comment]
(also squashed)

File: sys/include/net/gnrc/ipv6/ext.h [hunk 12]
```diff
@@ -66,6 +66,33 @@ gnrc_pktsnip_t *gnrc_ipv6_ext_demux(gnrc_pktsnip_t *pkt, unsigned nh);
...
+ * @return  @p pkt with all extension headers marked until the first
+ *          non-extension header.
+ * @return  NULL, if packet was consumed by the extension header handling.
+ * @return  NULL, on error. @p pkt is released with EINVAL in that case.
+ */
+gnrc_pktsnip_t *gnrc_ipv6_ext_process_all(gnrc_pktsnip_t *pkt,
+                                          uint8_t *nh);
+#else   /* defined(MODULE_GNRC_IPV6_EXT) || defined(DOXYGEN) */
+/* NOPs to make the usage code more readable */
+#define gnrc_ipv6_ext_process_all(pkt, first_nh)    (pkt);
```
Comment (jia200x) [inline sys/include/net/gnrc/ipv6/ext.h]: This is a little bit misleading IMO. This macro looks like a function and someone could spend some time looking to the actual implementation of `gnrc_ipv6_ext_process_all` instead of the NOP.

Also I would prefer to move this macro to the C file and just leave only the prototype here.

I would do something like:

[gnrc_ipv6_ext.c]
```
#ifdef MODULE_GNRC_IPV6_EXT
#define GNRC_IPV6_EXT_PROCESS_ALL(pkt, first_nh) gnrc_ipv6_ext_process_all(pkt, nh)
#else 
#define GNRC_IPV6_EXT_PROCESS_ALL(pkt, first_nh) (pkt)
#endif
```

That way, it's clear GNRC_IPV6_EXT_PROCESS_ALL is a macro

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: This is non-sensical. Then the user needs to use ifdefs anyway. The point of these NOP-macros is so you can use the functions in your code without worrying that the module is included or not.

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: But your proposal uses NOPs as well. It just adds another level of indirection.

Comment (jia200x) [reply sys/include/net/gnrc/ipv6/ext.h]: To be more accurate, I meant I'm not convinced with mixing a macro and a function with the same name. I would prefer just use a second macro (as I described) or a weak-like symbol mechanism.

Or at least add a tiny note before the usage of this function indicating it resolves to `pkt` if `MODULE_GNRC_IPV6_EXT` is not included :)

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: This mechanism is already used through-out the code and the behavior is the same as if `gnrc_ipv6_ext` would be in (but since the stack doesn't know about extension headers it is not handling them).

Comment (jia200x) [reply sys/include/net/gnrc/ipv6/ext.h]: ok, let's get it in

Comment (miri64) [reply sys/include/net/gnrc/ipv6/ext.h]: Some examples:

https://github.com/RIOT-OS/RIOT/blob/b709e63581a85a504b1020a025ef6f0e73f80b1d/sys/include/net/gnrc/icmpv6/error.h#L95-L102

https://github.com/RIOT-OS/RIOT/blob/b709e63581a85a504b1020a025ef6f0e73f80b1d/sys/include/net/gnrc/netif/internal.h#L470-L471

The ironic thing: Some reviewers specifically asked me to do it this way...

Comment (jia200x) [reply sys/include/net/gnrc/ipv6/ext.h]: yes, I see. There are several ways to handle stubs. I'm fine ;)

**Comment by jia200x** [comment]
Sending one HopByHop header:
```
>>> sendp(Ether() / 
...:       IPv6(dst=DST_ADDR) / 
...:       IPv6ExtHdrHopByHop() / 
...:       UDP(sport=1339, dport=1337) / 
...:       "\x00\x01", iface="tapbr0") 
```
I receive:
```udp server start 1337
Success: started UDP server on port 1337
> ipv6: GNRC_NETAPI_MSG_TYPE_RCV received
ipv6: Received (src = fe80::a9f1:fcc1:4389:4748, dst = fe80::4c3a:39ff:fe4f:45d2, next header = 0, length = 18)
ipv6_ext: next header = 0ipv6_ext: skipping over unsupported extension header
ipv6: forward nh = 17 to other threads
ipv6: waiting for incoming message.
PKTDUMP: data received:
~~ SNIP  0 - size:   2 byte, type: NETTYPE_UNDEF (0)
00000000  00  01
~~ SNIP  1 - size:   8 byte, type: NETTYPE_UDP (4)
   src-port:  1339  dst-port:  1337
   length: 10  cksum: 0xfc82
~~ SNIP  2 - size:   8 byte, type: NETTYPE_UNKNOWN (2)
00000000  11  00  01  04  00  00  00  00
~~ SNIP  3 - size:  40 byte, type: NETTYPE_IPV6 (1)
traffic class: 0x00 (ECN: 0x0, DSCP: 0x00)
flow label: 0x00000
length: 18  next header: 0  hop limit: 64
source address: fe80::a9f1:fcc1:4389:4748
destination address: fe80::4c3a:39ff:fe4f:45d2
~~ SNIP  4 - size:  20 byte, type: NETTYPE_NETIF (-1)
if_pid: 6  rssi: 0  lqi: 0
flags: 0x0
src_l2addr: 00:50:B6:8C:B5:1A
dst_l2addr: FF:FF:FF:FF:FF:FF
~~ PKT    -  5 snips, total size:  78 byte
```

When sending a malformed packet:
```
>>> sendp(Ether() / 
...:       IPv6(dst=DST_ADDR) / 
...:       IPv6ExtHdrHopByHop() / 
...:       IPv6ExtHdrHopByHop() / 
...:       UDP(sport=1339, dport=1337) / 
...:       "\x00\x01", iface="tapbr0")                                                                                                                        
WARNING: Mac address to reach destination 
```
I get
```
> ipv6: GNRC_NETAPI_MSG_TYPE_RCV received
ipv6: Received (src = fe80::a9f1:fcc1:4389:4748, dst = fe80::4c3a:39ff:fe4f:45d2, next header = 0, length = 26)
ipv6_ext: next header = 0ipv6_ext: skipping over unsupported extension header
ipv6: duplicate Hop-by-Hop header
ipv6: packet's extension header was errorneous or packet was consumed due to it
ipv6: waiting for incoming message.
```

As expected

**Comment by jia200x** [comment]
nothing interesting to report code-style-wise

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 13]
```diff
@@ -28,6 +30,124 @@
...
+ * problem message (if @ref net_gnrc_icmpv6_error is available), releases the
+ * packet with error `EINVAL`, and returns `false`.
+ *
+ * @note    This function should only be called *after* the first hop-by-hop
+ *          option header in the packet was parsed (if it exists) or if the
+ *          first extension header wasn't a hop-by-hop option header.
+ *
+ * @param[in] pkt       A packet in receive order to be checked with the current
+ *                      extension header (of type @p protnum) as its payload.
+ * @param[in] protnum   The protocol number of the header ingnrc_pktsnip_t::data
```
Comment (jia200x) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: missing space here :(

`ingnrc_pktsnip_t` => `in gnrc_pktsnip_t`

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 14]
```diff
@@ -28,6 +30,124 @@
...
+ *                      of @p pkt.
+ *
+ * @return  true, when @p protnum != @ref PROTNUM_IPV6_EXT_HOPOPT.
```
Comment (jia200x) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: AFAIS, the function return `true` when `protnum == PROTNUM_IPV6_EXT_HOPOPT`

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: Makes also more sense with the name ;-). Fixed.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c (hunk 13 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: Fixed.

**Comment by miri64** [comment]
> (that is already present in master, but no harm piggy-backing it here, right)

(Well it does if Git resolves the conflict automatically but wrong. Fixed that)

**Comment by jia200x** [comment]
ACK. Thanks!
